examples/autoignition-mpi.py

```diff
@@ -244,67 +298,144 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
-
-    if not check_step(current_step, nviz):  # If final step not an output step
-        if rank == 0:
-            logger.info("Checkpointing final state ...")
-        my_checkpoint(current_step, t=current_t,
-                      dt=(current_t - checkpoint_t),
-                      state=current_state)
-
-    if current_t - t_final < 0:
-        error_state = True
-
-    if error_state:
-        raise ValueError("Simulation did not complete successfully.")
+    def my_post_step(step, t, dt, state):
+        # Logmgr needs to know about EOS, dt, dim?
+        # imo this is a design/scope flaw
+        if logmgr:
+            set_dt(logmgr, dt)
+            set_sim_state(logmgr, dim, state, eos)
+            logmgr.tick_after()
+        return state, dt
+
+    def my_pre_step(step, t, dt, state):
+        dv = None
+        pre_step_errors = False
+
+        if logmgr:
+            logmgr.tick_before()
+
+        from mirgecom.simutil import check_step
+        do_viz = check_step(step=step, interval=nviz)
+        do_restart = check_step(step=step, interval=nrestart)
+        do_health = check_step(step=step, interval=nhealth)
+
+        if step == rst_step:  # don't do viz or restart @ restart
+            do_viz = False
+            do_restart = False
+
+        if do_health:
+            dv = eos.dependent_vars(state)
+            local_health_error = my_health_check(dv, dt)
+            health_errors = False
+            if comm is not None:
+                health_errors = comm.allreduce(local_health_error, op=MPI.LOR)
+            if health_errors and rank == 0:
+                logger.info("Fluid solution failed health check.")
+            pre_step_errors = pre_step_errors or health_errors
+
+        if do_restart:
+            my_write_restart(state, step, t)
+
+        if do_viz:
+            production_rates = eos.get_production_rates(state)
+            if dv is None:
+                dv = eos.dependent_vars(state)
+            my_write_viz(cv=state, dv=dv, step=step, t=t,
+                         production_rates=production_rates)
+
+        if pre_step_errors:
+            my_graceful_exit(cv=state, step=step, t=t,
+                             do_viz=(not do_viz), do_restart=(not do_restart),
+                             message="Error detected at prestep, exiting.")
```

Comment:
Exiting right away does not accomplish the desired behavior.  We want these visualizations and we don't want to overwrite them if they were already produced.

Reply:
?

By "exit right away", I'm saying you do the health check, then if it fails do the error vis/restart dump, then exit. As opposed the current approach which does the health check, then continues through the rest of the callback as normal, then finally writes the error vis/restart dump, but only if it hasn't been done already. This is the part that seems clunkly to me.

Reply:
I understand where you are coming from - but my preference is to check for the error only once.  I don't want to have exit logic at every point where an error can occur inside the callback.  I want to be able to make errors at any point inside the callback, and then check at the end if an error occurred.

Reply:
How about: https://github.com/majosm/mirgecom/commit/d106316b5fd96cfe0ddd0e7418aa61bc9dca015c

Reply:
Believe it or not, I like that just fine, and this I admit this is objectively "better code".  

That said, imo exceptions don't bring anything to the table here and make the coding and logic harder _for me_.  I will get used to it if we make it so, and I sense it is your strong preference to code it this way -  but just as a practical matter, the driver is in user-space, and I have the feeling that user-side folks (like MikeA) will also prefer the more familiar construction without exceptions.

Reply:
> Believe it or not, I like that just fine, and this I admit this is objectively "better code".
> 
> That said, imo exceptions don't bring anything to the table here and make the coding and logic harder _for me_. I will get used to it if we make it so, and I sense it is your strong preference to code it this way - but just as a practical matter, the driver is in user-space, and I have the feeling that user-side folks (like MikeA) will also prefer the more familiar construction without exceptions.

To try to give some justification for preferring this type of code structure...

Generally it's not safe for a sequence of execution to continue after an error has occurred. Say, for example, you have a function that transfers money between two bank accounts. Here are two ways you could write it:

Version 1 (break immediately):
```python
def transfer_money(old_account, new_account, amount):
    try:
        old_account.decrease_balance(amount)  # raises InsufficientBalanceError
        new_account.increase_balance(amount)

    except InsufficientBalanceError:
        raise TransferError("Failed to transfer; insufficient balance.")
```

Version 2 (wait until the end):
```python
def transfer_money(old_account, new_account, amount):
    transfer_error = False

    error = old_account.decrease_balance(amount)  # returns True
    transfer_error = transfer_error or error

    new_account.increase_balance(amount)  # oops, we created money

    if transfer_error:
        raise TransferError("Failed to transfer; insufficient balance.")
```

You can see that the latter version is incorrect. It's possible to fix it, but you would have to do something like wrap the subsequent steps in an `if not transfer_error:`, which is clunky (especially if you have multiple statements that can fail, since you'd need to wrap each separately).

Essentially all the `try`/`except` is doing here is letting you break out of the current execution when an error occurs. You could do the same thing with a function and early `return` (or a 1-iteration loop and `break` statements 🙂), this is just cleaner/less abusive.

Reply:
Right - but you've constructed a specific example where it is unsafe to continue on.   It is entirely possible, and equally easy, to construct a case and design a flow that gracefully handles the error or simply accumulates it, until handling it at the end - such as is already done in the current drivers.

Reply:
In fact, I might go so far as to say that your first example will leave the user's old account depleted of the amount they tried to transfer if adding it to the new account fails, and does nothing to differentiate between the different types of error that can occur.

It isn't immediately apparent to me how to handle the possible situations by throwing an exception at the site of the error, but if I accumulate the error and throw at the end, then it seems safer.

```python
def transfer_money(old_account, new_account, amount):
    insufficient_funds = old_account.decrease_balance(amount)
    if insufficient_funds:
       log("So sorry, insufficient funds to complete transaction.")
    transfer_failed = new_account.increase_balance(amount)
    if transfer_failed:
       log("So sorry, failed to add funds to new account.")
       recovery_failed = old_account.increase_balance(amount)
       if recovery_failed:
          trans_error = log_transaction_error(old_account, new_account, amount)
          log("Warning: Contact your financial institution re: {trans_error}")
    transfer_error = insufficient_funds or transfer_failed
    if transfer_error:
        raise TransferError("Your transaction could not be completed.")
```

Reply:
> Right - but you've constructed a specific example where it is unsafe to continue on. It is entirely possible, and equally easy, to construct a case and design a flow that gracefully handles the error or simply accumulates it, until handling it at the end - such as is already done in the current drivers.

The current code is already doing the `if not error:` thing; it's just cleverly disguised. 🙂 (The `not do_vis` etc. in the call to `graceful_exit`).

Reply:
> In fact, I might go so far as to say that your first example will leave the user's old account depleted of the amount they tried to transfer if adding it to the new account fails, and does nothing to differentiate between the different types of error that can occur.

The assumption in the example is that increasing the account balance is not something that's going to fail. If that's a possibility, you'd probably want a different API for the account object (maybe one that lets you pre-approve transactions before actually committing to them). But this is getting way outside the scope of what I was trying to illustrate.

> It isn't immediately apparent to me how to handle the possible situations by throwing an exception at the site of the error, but if I accumulate the error and throw at the end, then it seems safer.
> 
> ```python
> def transfer_money(old_account, new_account, amount):
>     transfer_error = False
>     insufficient_funds = old_account.decrease_balance(amount)
>     if insufficient_funds:
>        log("So sorry, insufficient funds to complete transaction.")
>        transfer_error = True
>     else:
>        transfer_failed = new_account.increase_balance(amount)
>        if transfer_failed:
>           transfer_error = True
>           log("So sorry, failed to add funds to new account.")
>           recovery_failed = old_account.increase_balance(amount)
>           if recovery_failed:
>              trans_error = log_transaction_error(old_account, new_account, amount)
>              log("Warning: Contact your financial institution re: {trans_error}")
>     if transfer_error:
>        raise TransferError("Your transaction could not be completed.")
> ```

Most of the time actions aren't reversible (may be my fault for choosing this as an example). It's better to not do something in the first place than to do it and subsequently attempt to undo it. If the recovery fails, you're in big trouble. 🙂 

~~Also, I think this still has a bug; it increases the balance in `new_account` even if `old_account.decrease_balance` fails.~~ I saw that ninja edit. 😛

Reply:
> ~Also, I think this still has a bug; it increases the balance in new_account even if old_account.decrease_balance fails.~ I saw that ninja edit. 😛

Right - but that's a further illustration of my point.  At least I, as a pedestrian, can inspect and understand what is happening and see that there's an error with the code and fix it.  Not so with the exceptions (or at least not as easily).  The code is more complicated to the pedestrian eye with exceptions, and thus more prone to my screwing the pooch with it.

The other point I tried to make is that as a simulation user, I may want to check for more things after the health check even if it fails, and then throw later.  I think that's fine - and can be done easily within the framework you propose. 

> The current code is already doing the if not error: thing; it's just cleverly disguised. 🙂 (The not do_vis etc. in the call to graceful_exit).
 
I don't think that's the same. This only prevents me from overwriting the previously written viz/restart with a new one.  After our most recent changes - this logic is no longer needed. It may have been misguided anyway.... at graceful exit, I really don't care about performance any more, so a duplicate write is fine.

Reply:
Here's the examples with your suggestion applied (ea98b59).